### ISSUE TITLE
Possible consensus bug with proposer slashings

### ISSUE THREAD

**Issue by JustinDrake** [sigp/lighthouse#1065]
The `process_proposer_slashings` function has two key steps, namely verifying the proposer slashings in parallel and then calling `slash_validator` once per `proposer_slashing`. The parallel proposer slashing verification seems to assume proposer slashings can be verified independently but there is some state dependency in `assert is_slashable_validator(proposer, get_current_epoch(state))` (see spec [here](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#proposer-slashings)).

Specifically, if the same (unslashed) validator is put several times in `process_proposer_slashings` then the second proposer slashing should throw an error. As far as I can tell `slash_validator` in Lighthouse will *not* throw an error if the validator is already slashed and `process_proposer_slashings` will not throw an error if the same validator is slashed several times.

**Comment by protolambda** [comment]
The spec has no explicit concurrency, Lighthouse may have optimized it wrong, and the edge case deserves a test. After first slashing, validator.slashed = True, and is-slashable will be False and abort the 2nd slashing.

**Comment by protolambda** [comment]
Can confirm the bug; https://github.com/sigp/lighthouse/blob/32074f0d095d901e77ce039ac02b3687f9cce76c/eth2/state_processing/src/per_block_processing.rs#L288

The operations are verified with a parallel iterator, only *reading* the state. Rust doesn't complain here. And then the slashings are applied, and the two proposer slashings that were valid on the pre-state, but not the specced intermediate state, would be confirmed.

Maybe it's a good idea to adjust `slash_validator` in the spec to explicitly check `is_slashable_validator`, so this can never happen again?

**Comment by paulhauner** [comment]
Thanks @JustinDrake, this is a great pickup.

I think this comes from back when I was working on [serenity-benches](https://github.com/sigp/serenity-benches), so I'll take responsibility for this.

We do this type of parallel execution for several block-operations processing functions because we saw a significant gain from doing BLS verification on multiple cores. Now we're doing batch BLS verification and we never* actually verify signatures during `per_block_processing`/`per_state_processing` I think we can realistically remove all these `par_iter`. @michaelsproul keen to hear your thoughts, as always.

*: we still verify signatures during deposits processing. Perhaps keeping _only_ concurrent signature verification here is sensible.

**Comment by michaelsproul** [comment]
I had a look at attestation processing, and it looks like it's safe because it only verifies the indexed attestations in parallel, and does the actual slashing in series, checking the non-empty slashing condition here:

https://github.com/sigp/lighthouse/blob/26bdc2927b6f51fd7c0deb909ea4e54c03087c45/eth2/state_processing/src/per_block_processing/verify_attester_slashing.rs#L97

For proposer slashings, I agree we could remove the parallel iterator entirely. And I guess we could defensively remove all of them?

----

Oh damn, and I just checked the op pool, which I thought handled this correctly, but it looks like it doesn't :\ It takes great care to avoid including duplicate attester slashings, but just piles in all the proposer slashings that are compatible with a single state:

https://github.com/sigp/lighthouse/blob/15a3af89662c4f2bc02f3eeaffca46be70a532b1/eth2/operation_pool/src/lib.rs#L237-L246

**Comment by JustinDrake** [comment]
> I think we can realistically remove all these `par_iter`

In this particular instance it may be wise to remove the `par_iter` especially if the performance gains are small. Having said that, as more of a meta comment for the codebase as a whole, I quite like that you guys are going "beyond" the spec to squeeze out performance :)

**Comment by JustinDrake** [comment]
As a meta question for @zedt3ster, why didn't the fuzzing infrastructure catch this bug or the recent consensus bug in Prysm regarding reward calculations?

**Comment by protolambda** [comment]
@JustinDrake Prysm introduced the bug just a few days before testnet launch with a last-minute v0.11 update. Can't catch something with fuzzing when the fuzzer doesn't run the code. But I'm sure it could have it were up to date.
For lighthouse, it's more questionable. Maybe the double proposer slashing is too specific to find with the fuzzer. The fuzzer would need to figure out to duplicate the exact slashing operation, not impossible though.

**Comment by zedt3ster** [comment]
So our beacon fuzz fuzzers were targeting `v0.10` up until a few days ago. As mentioned by @protolambda, the code that was instrumented didn't include the bug in question. We've also ran into multiple issues with the Prysm build (that are now resolved) which prevented us from integrating them into our continuous fuzzing.

We're currently revamping the differential fuzzer completely to include new mutation engines and leveraging structural fuzzing. In fact, we just identified an [SSZ decoding/processing bug](https://github.com/prysmaticlabs/prysm/issues/5658) in Prysm. The differential part of the fuzzers is currently "_on hold",_ we'll be sharing a new design this week - this type of bug could definitely be picked up in the future.
